python/pyspark/sql/connect/client/core.py

```diff
@@ -544,6 +544,25 @@ def fromProto(cls, pb: pb2.ConfigResponse) -> "ConfigResult":
         )
 
 
+class ForbidRecursion:
```

Comment:
I faced the same problem before too. In my case, I worked around by https://github.com/apache/spark/pull/43965/files#diff-831a8c82df3f07cbdaba03aaf7a0e9abaaf5dd6c63f9dd121e4a263e3094844eR1528.

Basically tries to get the config once, and do not retry if it fails. But wasn't sure if that's the best approach. cc  @heyihong

Reply:
I looked into https://github.com/apache/spark/pull/43965/files#diff-831a8c82df3f07cbdaba03aaf7a0e9abaaf5dd6c63f9dd121e4a263e3094844eR1528, I think this is not enough.

```try: except:``` guard you have is a good idea, however it won't be triggered immediately by recursion, and the code will walk again into RuntimeConf.get and got into infinite recursion again.

Reply:
There may be some simpler approach to deal with recursive error handling. Using ForbidRecursion seems to be a big hammer

Reply:
I like this hammer since it's very specific and allows to keep all existing fancies in error handling we already have :).

Regarding testing, I tried to write a test with mock stub which would fail, but I found that I need somewhat sophisticated GrpcError instance to pass this conversion https://github.com/apache/spark/blob/4398bb5d37328e2f3594302d98f98803a379a2e9/python/pyspark/sql/connect/client/core.py#L1559

Reply:
IMO, the logic that determines whether to display stack trace should be implemented on the sever side

Reply:
It seems it's controlled by server, we only need some configuration to do this.

Another approach I see is to write `spark.sql.connect.serverStacktrace.enabled` in some form of lazy fetch and also put something forbidding recursion into this loader instead

Reply:
We probably don't need two extra config request round trip to know whether to display stack trace or not. We can just determine whether to display stack trace if stack trace field in the response is not empty

Reply:
That's a good idea, let's try this